clang/lib/CIR/CodeGen/CIRGenBuiltin.cpp

```diff
@@ -642,6 +660,48 @@ RValue CIRGenFunction::buildBuiltinExpr(const GlobalDecl GD, unsigned BuiltinID,
       Result = builder.createBoolToInt(Result, ResultType);
     return RValue::get(Result);
   }
+
+  case Builtin::BIalloca:
+  case Builtin::BI_alloca:
+  case Builtin::BI__builtin_alloca_uninitialized:
+  case Builtin::BI__builtin_alloca: {
+    // Get alloca size input
+    mlir::Value AllocaSize = buildScalarExpr(E->getArg(0));
+
+    // The alignment of the alloca should correspond to __BIGGEST_ALIGNMENT__.
+    const TargetInfo &TI = getContext().getTargetInfo();
+    const CharUnits AlignmentInBytes =
+        getContext().toCharUnitsFromBits(TI.getSuitableAlign());
+
+    auto AlignIntAttr = CGM.getSize(AlignmentInBytes);
+
+    // Emit the alloca op with type `u8 *` to match the semantics of
+    // `llvm.alloca`. We later bitcast the type to `void *` to match the
+    // semantics of C/C++
+    // FIXME: It may make sense to allow AllocaOp of type `u8` to return a
+    // pointer of type `void *`. This will require a change to the allocaOp
+    // verifier.
+    auto Alloca = builder.create<mlir::cir::AllocaOp>(
```

Comment:
CIRGenBuilder should have its own `createAlloca`, like it does for other operations. We usually do a best effort to add those (see both `CIRGenBuilder.h` and `CIRBaseBuilder.h`) as we go, and looks like this is a good moment to add this one (even though we missed some places in the past). CIRGen usually uses buildAlloca, but that expectes CIRGen specific stuff, we also need a createAlloca to abstract the actual materialization of `builder.create`.

Reply:
I would do this happily! 
Can you help me understand the underlying philosophy/design behind it? If there's something I can read it would be even better than you reciting it for me 😇  
In addition, I'd be happy for a clarification on this comment:

> we also need a createAlloca to abstract the actual materialization of `builder.create`.

`CIRGenBuiltin.cpp` is full of direct calls to `builder.create<>()`. Is it just "legacy" code expecting to be properly wrapped over time? :)

Reply:
And another note: seems like this isn't implemented in clang's derived `CGBuilderTy`, but only in the base `IRBuilderBase` class. I guess this is not applicable in MLIR since the operations are derived from the various dialects. This fact raises even more questions for me on the choice to hide the create materialization (as it isn't hidden for SCF ops I've sampled, for example 😇 ). I think I'd stop wondering about it and wait for your answer! 🙏

Reply:
> Can you help me understand the underlying philosophy/design behind it

It's easier to call into the helper function than having to figure out which version of Op::build to use most of the times. LLVM also has these convenient helpers and are a good source of inspiration.

> Is it just "legacy" code expecting to be properly wrapped over time?

Yes.

> And another note: seems like this isn't implemented in clang's derived CGBuilderTy, but only in the base IRBuilderBase class

We have things in both `CIRGenBuilder.h` and `CIRBaseBuilder.h` depending on whether the helpers are free of CIRGen specific params or not. Right now we have things in `CIRGenBuilder.h` that should be moved to  `CIRBaseBuilder.h`, we are currently moving those as we see need, but PRs to tackle that are also welcome.